### PR TITLE
Rdm 4374 - Validate and reflect call back changes on the front end DynamicList

### PR DESCRIPTION
**Pull request by mohammedanas**
https://tools.hmcts.net/jira/browse/RDM-4374

Currently we have the 'FixedList' base type which services use to configure a list of options that the user can pick from. This data is fixed as per the configuration.

With the introduction of 'DynamicList' base type services will be able to update the list of options in the list dynamically via the AboutToStart and/or MidEvent call backs

This story is to create the 'DynamicList' base type and add to the definition store

Store selected key and selected value

    Store selected key valued pair
    Store options key value pairs
    Max validation same as FixedList

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseEvent.java [hunk 1]
```diff
@@ -103,7 +103,7 @@ public void setPostState(String postState) {
         this.postState = postState;
     }
 
-    public String getCallBackURLAboutToStartEvent() {
+    public String   getCallBackURLAboutToStartEvent() {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseEvent.java]: can you remove this space please

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/DynamicListItem.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package uk.gov.hmcts.ccd.domain.model.definition;
+
+public class DynamicListItem extends FixedListItem{
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/DynamicListItem.java]: missing space FixedListItem{

File: src/main/java/uk/gov/hmcts/ccd/domain/types/DynamicListValidator.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.Collections;
+import java.util.List;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import org.apache.commons.lang3.StringUtils;
+import uk.gov.hmcts.ccd.domain.model.definition.CaseField;
+
+@Named
+@Singleton
+public class DynamicListValidator implements BaseTypeValidator {
```
[the remark this answers is not in the document]
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/types/DynamicListValidator.java]: I initially tried to reuse the existing validator, but the way the contract as been agreed with I needed a different validator, with fixed list the format comes from definition with dynamic it comes from service in the following format. 
        "dynamicList": {
          "default": {
            "code": "FixedList1",
            "label": "Fixed List 1"
          },
          "dynamic_list_items": [{
            "code": "FixedList1",
            "label": "Fixed List 1"
          }, {
            "code": "FixedList2",
            "label": "Fixed List 2"
          }, {
            "code": "FixedList3",
            "label": "Fixed List 3"
          }, {
            "code": "FixedList4",
            "label": "Fixed List 4"
          }, {
            "code": "FixedList5",
            "label": "Fixed List 5"
          }, {
            "code": "FixedList6",
            "label": "Fixed List 6"
          }, {
            "code": "FixedList7",
            "label": "Fixed List 7"
          }
          ]
        }

[the remark this answers is not in the document]
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/types/DynamicListValidator.java]: Because the format is different from normal fixed_list_items I felt it was safer to have a dynamic_list_items plus changing name across of fixed_list_items to list_items which seemed to be like a lot bigger change UI and regression testing across.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java [hunk 4]
```diff
@@ -40,10 +50,29 @@ public CaseViewField build(CaseField caseField, CaseEventField eventField) {
     public CaseViewField build(CaseField caseField, CaseEventField eventField, Object value) {
         final CaseViewField field = build(caseField, eventField);
         field.setValue(value);
-
+        getFieldType(field, value);
         return field;
     }
 
+    private void getFieldType(CaseViewField caseField, Object value) {
```
[the remark this answers is not in the document]
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: done

[the remark this answers is not in the document]
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: done

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java [hunk 5]
```diff
@@ -40,10 +50,29 @@ public CaseViewField build(CaseField caseField, CaseEventField eventField) {
...
+
+        if (caseField.getFieldType().getType().equals(DYNAMIC_LIST) && value != null) {
+            caseField.setValue(((ObjectNode) value).get(DEFAULT));
+            caseField.getFieldType().setDynamicListItems(processDynamicList((ObjectNode) value));
+        }
+    }
+
+    private List<DynamicListItem> processDynamicList(ObjectNode value) {
```
[the remark this answers is not in the document]
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: Because part of the requirement was to add list of items as part of value to be saved along with value.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseEvent.java (hunk 1 above)
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseEvent.java]: done

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/DynamicListItem.java (hunk 2 above)
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/DynamicListItem.java]: removed this class now as per discussion

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/FieldType.java [hunk 6]
```diff
@@ -27,6 +27,8 @@
     private String regularExpression = null;
     @JsonProperty("fixed_list_items")
     private List<FixedListItem> fixedListItems = new ArrayList<>();
+    @JsonProperty("dynamic_list_items")
```
[the remark this answers is not in the document]
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/FieldType.java]: got rid of dynamic_list_items property

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/FieldType.java [hunk 7]
```diff
@@ -72,6 +72,7 @@ public void setFixedListItems(List<FixedListItem> fixedListItems) {
         this.fixedListItems = fixedListItems;
     }
 
+
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/FieldType.java]: can u remove this please

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java [hunk 8]
```diff
@@ -53,4 +63,23 @@ public CaseViewField build(CaseField caseField, CaseEventField eventField, Objec
...
+
+        if (caseField.getFieldType().getType().equals(DYNAMIC_LIST) && value != null) {
+            caseField.setValue(((ObjectNode) value).get(DEFAULT));
+            caseField.getFieldType().setFixedListItems(processDynamicList((ObjectNode) value));
+        }
+    }
+
+    private List<FixedListItem> processDynamicList(ObjectNode value) {
+        List<FixedListItem> result = new ArrayList<>();
+        value.get(DYNAMIC_LIST_ITEMS).elements().forEachRemaining(dynamicList -> {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: I thought we said we would use `fixed_list_items`

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java [hunk 9]
```diff
@@ -53,4 +63,23 @@ public CaseViewField build(CaseField caseField, CaseEventField eventField, Objec
             .map(eventField -> build(caseFieldMap.get(eventField.getCaseFieldId()), eventField, data != null ? data.get(eventField.getCaseFieldId()) : null))
             .collect(Collectors.toList());
     }
+
+    private void getFieldTypeWithDynamicListsPopulated(CaseViewField caseField, Object value) {
+
+        if (caseField.getFieldType().getType().equals(DYNAMIC_LIST) && value != null) {
+            caseField.setValue(((ObjectNode) value).get(DEFAULT));
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: what happens when a user selected on the ui a valued for the dynamic list and we stored that value. Would we be overriding the selected value with the default here?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java [hunk 10]
```diff
@@ -40,7 +50,7 @@ public CaseViewField build(CaseField caseField, CaseEventField eventField) {
     public CaseViewField build(CaseField caseField, CaseEventField eventField, Object value) {
         final CaseViewField field = build(caseField, eventField);
         field.setValue(value);
-
+        getFieldTypeWithDynamicListsPopulated(field, value);
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: can we rename to `populateDynamicList`

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java [hunk 11]
```diff
@@ -53,4 +63,23 @@ public CaseViewField build(CaseField caseField, CaseEventField eventField, Objec
...
+        if (caseField.getFieldType().getType().equals(DYNAMIC_LIST) && value != null) {
+            caseField.setValue(((ObjectNode) value).get(DEFAULT));
+            caseField.getFieldType().setFixedListItems(processDynamicList((ObjectNode) value));
+        }
+    }
+
+    private List<FixedListItem> processDynamicList(ObjectNode value) {
+        List<FixedListItem> result = new ArrayList<>();
+        value.get(DYNAMIC_LIST_ITEMS).elements().forEachRemaining(dynamicList -> {
+            FixedListItem listItem = new FixedListItem();
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: How difficult would be to rename `FixedListItem` to `ListItem`? Don't think we risk breaking anything

File: src/main/java/uk/gov/hmcts/ccd/domain/types/DynamicListValidator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/types/DynamicListValidator.java]: ok got it 👍

File: src/main/java/uk/gov/hmcts/ccd/domain/types/DynamicListValidator.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    private static final String DEFAULT = "default";
+    private static final String LABEL = "label";
+
+    @Override
+    public BaseType getType() {
+        return BaseType.get(TYPE_ID);
+    }
+
+    @Override
+    public List<ValidationResult> validate(String dataFieldId, JsonNode dataValue, CaseField caseFieldDefinition) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/types/DynamicListValidator.java]: beware if you look at FixedListValidator there are some validation checks there that we don't have here. For example https://github.com/hmcts/ccd-data-store-api/blob/5f9adffe7b15c8e81273a275f1bf6def3d6a76ca/src/main/java/uk/gov/hmcts/ccd/domain/types/FixedListValidator.java#L35 
but others too?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java (hunk 9 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: related to this, where do we read the current selected value of the dynamic list?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java [hunk 13]
```diff
@@ -1,21 +1,34 @@
 package uk.gov.hmcts.ccd.domain.model.aggregated;
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: we said this code is no longer needed. Is that right? Can we remove it then?

File: src/test/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilderTest.java [hunk 14]
```diff
@@ -19,6 +19,7 @@
 import static uk.gov.hmcts.ccd.domain.service.common.TestBuildersUtil.FieldTypeBuilder.aFieldType;
```
Comment (mario-paniccia) [inline src/test/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilderTest.java]: is this still needed?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java [hunk 15]
```diff
@@ -12,6 +12,7 @@
 
 import static java.util.Optional.ofNullable;
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: revert these changes please

File: src/test/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilderTest.java (hunk 14 above)
Comment (mohammedanas) [reply src/test/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilderTest.java]: reverted

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java (hunk 15 above)
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: reverted

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java (hunk 13 above)
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: reverted

File: src/main/java/uk/gov/hmcts/ccd/domain/types/DynamicListValidator.java (hunk 12 above)
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/types/DynamicListValidator.java]: Spoke to @amansinghb and there would be a separate ticket dealing with validations comming soon

File: src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java (hunk 11 above)
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/aggregated/CaseViewFieldBuilder.java]: Should leave we this when we will be standarding fixed-lists to lists across the applications?
